pyfr/integrators/dual/base.py

```diff
@@ -29,3 +121,97 @@ def _dual_time_source(self):
     @abstractmethod
     def finalise_step(self, currsoln):
         pass
+
+    @memoize
+    def prolrest(self, l1, l2):
+        prolrestkerns = proxylist([])
+
+        for etype in self.system.ele_types:
+            l1sys = self.dualsystems[l1]
+            l2sys = self.dualsystems[l2]
+
+            l1eles = l1sys.ele_map[etype]
+            l2eles = l2sys.ele_map[etype]
+
+            prolrestkerns.append(
+                self.backend.kernel('mul',
+                                    self._get_prolrestmatrix(l1eles, l2eles),
+                                    l1eles.scal_upts_inb,
+                                    out=l2eles.scal_upts_inb)
+            )
+
+        return prolrestkerns
+
+    def _get_prolrestmatrix(self, l1eles, l2eles):
+        ub1 = l1eles.basis.ubasis
+        l2b = l2eles.basis
+
+        mat = ub1.nodal_basis_at(l2b.upts)
+        return self.backend.const_matrix(mat, tags={'align'})
+
+    def restrict(self, l1, l2, dt):
+        sys, idxcurr = self.dualsystems, self._dualidxcurr
+        regidx, mgregidx = self._dualregidx, self._mg_regidx
+
+        add, rhs = self._add, self._rhs_with_dts
+
+        # Registers [level][(stepper_regs)(source_regs)(multigrid_regs)]
+        # R = [l(1)][mg(1)] = -∇·f - dQ/dt
+        rhs(self.tcurr, idxcurr[l1], regidx[l1][mgregidx[1]],
+            c=1/dt, level=l1, passmg=True)
+
+        l = 0 if l1 == 0 else 1
+        # d = [l(1)][mg(2)] = r - R
+        add(-1, regidx[l1][mgregidx[1]], l, regidx[l1][mgregidx[0]], level=l1)
```

Comment:
This indexing is sufficiently messy that it is likely we're abstracting the problem incorrectly.